bigquery/storage/managedwriter/routers_test.go

```diff
@@ -0,0 +1,351 @@
+			router:              newSimpleRouter(""),
+			numWriters:          1,
+			numMultiplexWriters: 1,
+		},
+		{
+			desc:                "SimpleRouter",
+			router:              newSimpleRouter(""),
+			numWriters:          10,
+			numMultiplexWriters: 10,
+		},
+		{
+			desc:                "SharedRouterNoMultiplex",
+			router:              newSharedRouter(false, 0),
+			numWriters:          1,
+			numMultiplexWriters: 1,
+		},
+		{
+			desc:                "SharedRouterNoMultiplex",
+			router:              newSharedRouter(false, 0),
+			numWriters:          10,
+			numMultiplexWriters: 10,
+		},
+		{
+			desc:                "SharedRouterMultiplex1conn",
+			router:              newSharedRouter(true, 1),
+			numWriters:          1,
+			numMultiplexWriters: 1,
+		},
+		{
+			desc:                "SharedRouterMultiplex1conn",
+			router:              newSharedRouter(true, 1),
+			numWriters:          10,
+			numMultiplexWriters: 10,
+		},
+		{
+			desc:                "SharedRouterMultiplex1conn",
+			router:              newSharedRouter(true, 1),
+			numWriters:          50,
+			numMultiplexWriters: 50,
+		},
+		{
+			desc:                "SharedRouterMultiplex10conn",
+			router:              newSharedRouter(true, 10),
+			numWriters:          50,
+			numMultiplexWriters: 50,
+		},
+	} {
+
+		ctx, cancel := context.WithCancel(context.Background())
+		pool := &connectionPool{
+			ctx:    ctx,
+			cancel: cancel,
+			open: func(opts ...gax.CallOption) (storagepb.BigQueryWrite_AppendRowsClient, error) {
+				return &testAppendRowsClient{}, nil
+			},
+		}
+		if err := pool.activateRouter(bm.router); err != nil {
+			b.Errorf("%q: activateRouter: %v", bm.desc, err)
+		}
+
+		// setup both explicit and default stream writers.
+		var writers []*ManagedStream
+		for i := 0; i < bm.numWriters; i++ {
+			wCtx, wCancel := context.WithCancel(ctx)
+			writer := &ManagedStream{
+				id:             newUUID(writerIDPrefix),
+				streamSettings: &streamSettings{streamID: "projects/foo/datasets/bar/tables/baz/streams/abc123"},
+				ctx:            wCtx,
+				cancel:         wCancel,
+				retry:          newStatelessRetryer(),
+			}
+			writers = append(writers, writer)
+		}
+		for i := 0; i < bm.numMultiplexWriters; i++ {
+			wCtx, wCancel := context.WithCancel(ctx)
+			writer := &ManagedStream{
+				id:             newUUID(writerIDPrefix),
+				streamSettings: &streamSettings{streamID: "projects/foo/datasets/bar/tables/baz/streams/_default"},
+
+				ctx:    wCtx,
+				cancel: wCancel,
+				retry:  newStatelessRetryer(),
+			}
+			writers = append(writers, writer)
+		}
+
+		for k, writer := range writers {
+			if err := pool.addWriter(writer); err != nil {
+				b.Errorf("addWriter %d: %v", k, err)
+			}
+		}
+		benchName := fmt.Sprintf("%s_%dexwriters_%dmpwriters", bm.desc, bm.numWriters, bm.numMultiplexWriters)
+		b.Run(benchName, func(b *testing.B) {
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				// route each writer once
+				writer := writers[0]
```

Comment:
i ?

Reply:
reworked the benchmarking so we align pickConnection to random writers each iteration.